src/topography/topography.jl

```diff
@@ -105,11 +107,107 @@ Surface elevation for 3D cosine hills.
 """
 function topography_cosine_3d(coord)
     FT = Geometry.float_type(coord)
-    (; x, y) = coord
-    (; h_max, λ) = cosine_params(FT)
+    x, y = coord.x, coord.y
+    cos_params = cosine_params(FT)
+    h_max, λ = cos_params.h_max, cos_params.λ
     return topography_cosine(x, y, λ, λ, h_max)
 end
 
 topography_cosine(x, y, λ_x, λ_y, h_max) =
     h_max * cospi(2 * x / λ_x) * cospi(2 * y / λ_y)
 cosine_params(::Type{FT}) where {FT} = (; h_max = FT(25), λ = FT(25e3))
+
+abstract type AbstractTopography end
+
+topo_func(t::AbstractTopography) = error("No topography function for topography $t")
+
+struct NoTopography <: AbstractTopography end
+
+# Analytical topography types for idealized test cases
+
+"""
+    CosineTopography(; h_max = 25, λ = 25e3, dim = 2)
+
+Cosine hill topography in 2D or 3D.
+
+# Arguments
+- `h_max::FT`: Maximum elevation (m)
+- `λ::FT`: Wavelength of the cosine hills (m)
+- `dim::Int`: Spatial dimension (2 or 3)
+"""
+Base.@kwdef struct CosineTopography{D, FT} <: AbstractTopography
+    h_max::FT = 25.0
+    λ::FT = 25e3
+end
+
+CosineTopography2D() = CosineTopography{2, Float64}()
+CosineTopography3D() = CosineTopography{3, Float64}()
+
+topo_func(t::CosineTopography{2}) = topography_cosine_2d
+topo_func(t::CosineTopography{3}) = topography_cosine_3d
+
+"""
+    AgnesiTopography(; h_max = 25, x_center = 50e3, a = 5e3)
+
+Witch of Agnesi mountain topography for 2D simulations.
+
+# Arguments
+- `h_max`: Maximum elevation (m)
+- `x_center`: Center position (m)
+- `a`: Mountain width parameter (m)
+"""
+Base.@kwdef struct AgnesiTopography{FT} <: AbstractTopography
+    h_max::FT = 25.0
+    x_center::FT = 50e3
+    a::FT = 5e3
+end
+
+topo_func(::AgnesiTopography) = topography_agnesi
+
+
+"""
+    ScharTopography(; h_max = 25, x_center = 50e3, λ = 4e3, a = 5e3)
+
+Schar mountain topography for 2D simulations.
+
+# Arguments
+- `h_max`: Maximum elevation (m)
+- `x_center`: Center position (m)
+- `λ`: Wavelength parameter (m)
+- `a`: Mountain width parameter (m)
+"""
+Base.@kwdef struct ScharTopography{FT} <: AbstractTopography
+    h_max::FT = 25.0
+    x_center::FT = 50e3
+    λ::FT = 4e3
+    a::FT = 5e3
+end
+topo_func(::ScharTopography) = topography_schar
```

Comment:
```suggestion
topography_function((; h_max, x_center, λ, a)::ScharTopography) =
    (; x) -> h_max * exp(-(x - x_center)^2 / a^2) * cospi((x - x_center) / λ)^2
```
The methods of `topo_func` can be combined with the definitions of `topography_<...>`, using dispatch like in `steady_state_velocity`. To avoid having a lot of closures, you could also write this as
```suggestion
topography_function((; h_max, x_center, λ, a)::ScharTopography, (; x)) =
    h_max * exp(-(x - x_center)^2 / a^2) * cospi((x - x_center) / λ)^2
topography_function(topo) = Base.Fix1(topography_function, topo)
```
The topography parameters are currently being duplicated, e.g., in `ScharTopography` and `schar_params(FT)`. It would be good to reduce duplication by removing `schar_params(FT)` and using the fields of `ScharTopography` instead, like in this example.

Reply:
Thanks, this is a much simpler pattern.